### PR TITLE
Wrong use of generics in ArrayComparator#Reverse

### PR DESCRIPTION
**Pull request by piddubnyi**
Solving https://github.com/jcabi/jcabi-immutable/issues/21
Changed generic usage to avoid wrong usage.

### CODE REVIEW

**Comment by dmarkov** [comment]
@piddubnyi I will find a reviewer for your pull requests shortly, thanks for contribution!

**Comment by dmarkov** [comment]
@pinaf could you please review this PR

**Comment by pinaf** [comment]
@piddubnyi let's add the proposed test case.

**Comment by pinaf** [comment]
@piddubnyi `<T extends Comparable<T>` seems recursive to me. perhaps you mean `<T extends Comparable>`?

File: src/main/java/com/jcabi/immutable/ArrayComparator.java [hunk 1]
```diff
@@ -50,7 +50,8 @@
      * @param <T> Type of argument
      */
     @Immutable
-    final class Default<T> implements ArrayComparator<T>, Serializable {
+    final class Default<T> implements ArrayComparator<T>,
```
Comment (pinaf) [inline src/main/java/com/jcabi/immutable/ArrayComparator.java]: @piddubnyi let's revert this change

**Comment by piddubnyi** [comment]
@pinaf <T extends Comparable<T>> is kind of recursive, but otherwise qulice complain about raw type, should I add annotation to ignore this check?

Also, the proposed test is just showing the possibility of getting ClassCastException, with new implementation it is just not compiles because of wrong generic types. Only positive test is possible, should I add it ?

**Comment by pinaf** [comment]
@piddubnyi you are correct, please do add the positive test. We'll see what @yegor256 says about the recursive generics use.

**Comment by piddubnyi** [comment]
@pinaf Please check if such test is enough.

File: src/test/java/com/jcabi/immutable/ArrayComparatorTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link ArrayComparator} implementations.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 2.0
```
Comment (pinaf) [inline src/test/java/com/jcabi/immutable/ArrayComparatorTest.java]: @piddubnyi let's go with 1.5 here

File: src/test/java/com/jcabi/immutable/ArrayComparatorTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class ArrayComparatorTest {
```
Comment (pinaf) [inline src/test/java/com/jcabi/immutable/ArrayComparatorTest.java]: @piddubnyi let's make this final

File: src/test/java/com/jcabi/immutable/ArrayComparatorTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * ReverseComparator can compare.
+     * @throws Exception If fails
+     */
+    @Test
+    public final void reverseComparatorCanCompare() throws Exception {
+        final ArrayComparator.Reverse<Integer> comparator =
+            new ArrayComparator.Reverse<Integer>();
+        final int result = comparator.compare(1, 2);
```
Comment (pinaf) [inline src/test/java/com/jcabi/immutable/ArrayComparatorTest.java]: @piddubnyi we can inline this variable

File: src/test/java/com/jcabi/immutable/ArrayComparatorTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class ArrayComparatorTest {
+
+    /**
+     * ReverseComparator can compare.
+     * @throws Exception If fails
+     */
+    @Test
+    public final void reverseComparatorCanCompare() throws Exception {
+        final ArrayComparator.Reverse<Integer> comparator =
```
Comment (pinaf) [inline src/test/java/com/jcabi/immutable/ArrayComparatorTest.java]: @piddubnyi we can inline this one also

File: src/test/java/com/jcabi/immutable/ArrayComparatorTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.immutable;
+
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Test case for {@link ArrayComparator} implementations.
```
Comment (pinaf) [inline src/test/java/com/jcabi/immutable/ArrayComparatorTest.java]: @piddubnyi let's remove the word `implementations`.

**Comment by pinaf** [comment]
@piddubnyi yes it is. please see a few more comments.

**Comment by piddubnyi** [comment]
@pinaf Fixed

**Comment by pinaf** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by dmarkov** [comment]
@pinaf **26 mins** was added to your account, many thanks for your contribution (payment AP-50N657616J286232A)! The task took 92 hours and 26 mins.. you're getting extra minutes here (c=11). +26 to your rating, your total score is [+6290](http://www.netbout.com/b/35128?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy now pls
